base/base/CTArray.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+
+/// Compile-time array that can be used as a template argument.
+template <class T, size_t N>
+struct CTArray
```

Comment:
1. You can't specialize std::array||std::array (and even if you could, it would be quite controversial).
2. A distinct type is better for semantics.
3. calling constexpr find on CTArray (contains) is simpler than calling an explicit find on std::array

Reply:
2. Why?
1. and 2. Why can't you just introduce `constexpr` functions for this purposes?

Reply:
2. CTArray is a distinct type suitable for using in template arguments.  std::array is a general-purpose-type. In user code it's better to separate types suitable for specialized tasks,
3. calling contains() on a constexpr object is more simple than declaring a constexpr object and making a constexpr it/bool as a find result. Member functions are prefered in STL over free functions, so I'll stick to that.

Reply:
Why do you need to use array as template argument in the first place?

Reply:
https://github.com/ClickHouse/ClickHouse/pull/29690/files#diff-327a0de53df2de18dff948ffd4b7fdae50243402505dac868c9884535fc0b12eR43

Main use is over here --the only way is to iterate over a container supplied as a template argument if you need compiler-time dispatching.

Reply:
I see situation as follows:
1. You want to process each element of the container in the compile time.
2. You need to call `static_for` on this container.
And here we've got a question. Why do you need to pass this container as a template parameter while you can pass it as an `static_for` argument. It'll still work. (And probably it'll compile faster because this solution will require less template instantiations)

Reply:
template argument gives an explicit "argument will be passed in compile-time", which is not always true for simply passing a constexpr argument into a constexpr function (as constexpr is not consteval). Imo current implementation is better for end users as it's easier to understand.

Reply:
Proposed changes are also consistent with current static_for implementation that accepts a range as template arguments, not regular arguments.

Reply:
And why we can't make it `consteval`?

Reply:
Probably in that case users should not call `static_for` at all.

Reply:
You say that static_for should be a more general solution and then say "users should not call ... at all", which does not look consistent. I've linked an example how other users can use static_for with template arguments to build higher abstractions (e.g. dispatchOverTypes) without copy-paste.

Reply:
`static_for` allows to perform loop in the **compile time**. If loop body (function passed to `static_for`) can't be evaluated in the compile time, it's completely OK to have compilation error, because it that case user should use `std::for_each`.

Reply:
static_for is essentially a switch-case statement (as taken from D). It's not OK to have a CE when you can achieve switch-case semantics.